### PR TITLE
Add Azure infrastructure provider

### LINKED ISSUE #23: Implement `Infrastructure` controller for Azure provider
Similar to how we have implemented the `Infrastructure` extension resource controller for the [AWS provider](https://github.com/gardener/gardener-extensions/tree/master/controllers/provider-aws/pkg/infrastructure) let's please now do it for Azure.

Based on [the current implementation](https://github.com/gardener/gardener/blob/master/example/90-shoot-azure.yaml#L13-L20) the `InfrastructureConfig` should look like this:

```yaml
apiVersion: azure.provider.extensions.gardener.cloud/v1alpha1
kind: InfrastructureConfig
resourceGroup:
  name: mygroup
networks:
  vnet: # specify either 'name' or 'cidr'
  # name: my-vnet
    cidr: 10.250.0.0/16
  workers: 10.250.0.0/19
```

Based on the [current implementation](https://github.com/gardener/gardener/blob/master/charts/seed-terraformer/charts/azure-infra/templates/_main.tf#L67-L93) the `InfrastructureStatus` should look like this:

```yaml
---
apiVersion: azure.provider.extensions.gardener.cloud/v1alpha1
kind: InfrastructureStatus
resourceGroup:
  name: mygroup
networks:
  vnet:
    name: my-vnet
  subnets:
  - purpose: nodes
    name: my-subnet
availabilitySets:
- purpose: nodes
  id: av-set-id
  name: av-set-name
routeTables:
- purpose: nodes
  name: route-table-name
securityGroups:
- purpose: nodes
  name: sec-group-name
```

The current infrastructure creation/deletion implementation can be found [here](https://github.com/gardener/gardener/blob/master/pkg/operation/cloudbotanist/azurebotanist/infrastructure.go). Please try to change as little as possible (with every change the risk that we break something increases!) and just move the code over into the extensions infrastructure actuator.

### PR DESCRIPTION
**Pull request by zanetworker**
**What this PR does / why we need it**:
Add Azure infrastructure provider

**Which issue(s) this PR fixes**:
Fixes #23 

```noteworthy user
The Azure provider does now have support for the `Infrastructure` extension.
```

### CODE REVIEW

File: controllers/provider-azure/pkg/apis/azure/v1alpha1/types_infrastructure.go [hunk 1]
```diff
@@ -0,0 +1,117 @@
+package v1alpha1
```
Comment (rfranzke) [inline controllers/provider-azure/pkg/apis/azure/v1alpha1/types_infrastructure.go]: license header

File: controllers/provider-azure/pkg/controller/infrastructure/actuator_delete.go [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+	if err != nil {
+		return err
+	}
+
+	tf, err := internal.NewTerraformer(a.restConfig, clientAuth, infrastructure.TerraformerPurpose, infra.Namespace, infra.Name)
+	if err != nil {
+		return err
+	}
+
+	configExists, err := tf.ConfigExists()
```
Comment (rfranzke) [inline controllers/provider-azure/pkg/controller/infrastructure/actuator_delete.go]: All this complexity with flow etc. is not needed for Azure. Just call `tf.Destroy` ([as it was in the Azure Botanist](https://github.com/gardener/gardener/blob/master/pkg/operation/cloudbotanist/azurebotanist/infrastructure.go#L76-L78)).

File: controllers/provider-azure/pkg/controller/infrastructure/actuator_reconcile.go [hunk 3]
```diff
@@ -0,0 +1,45 @@
+package infrastructure
```
Comment (rfranzke) [inline controllers/provider-azure/pkg/controller/infrastructure/actuator_reconcile.go]: license header

File: controllers/provider-azure/pkg/internal/infrastructure/infrastructure.go [hunk 4]
```diff
@@ -0,0 +1,15 @@
+package infrastructure
```
Comment (rfranzke) [inline controllers/provider-azure/pkg/internal/infrastructure/infrastructure.go]: license header

File: controllers/provider-azure/pkg/internal/infrastructure/terraform.go [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+		TerraformerOutputKeySubnetName,
+		TerraformerOutputKeyVNetName,
+	}
+
+	vars, err := tf.GetStateOutputVariables(outputKeys...)
+	if err != nil {
+		return nil, err
+	}
+
+	state := &TerraformState{
```
Comment (rfranzke) [inline controllers/provider-azure/pkg/internal/infrastructure/terraform.go]:
```suggestion
	return &TerraformState{
		...
	}, nil
```

File: controllers/provider-azure/pkg/internal/infrastructure/terraform.go [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+		SubnetName:          vars[TerraformerOutputKeySubnetName],
+	}
+	return state, nil
+}
+
+// StatusFromTerraformState computes an InfrastructureStatus from the given
+// Terraform variables.
+func StatusFromTerraformState(state *TerraformState) *azurev1alpha1.InfrastructureStatus {
+	var (
+		status = &azurev1alpha1.InfrastructureStatus{
```
Comment (rfranzke) [inline controllers/provider-azure/pkg/internal/infrastructure/terraform.go]: No need for a `var ()` block, return immediately the struct.

File: controllers/provider-azure/pkg/internal/terraform.go [hunk 7]
```diff
@@ -0,0 +1,45 @@
+package internal
```
Comment (rfranzke) [inline controllers/provider-azure/pkg/internal/terraform.go]: license header

File: controllers/provider-azure/pkg/internal/terraform.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+const (
+	// TerraformVarClientID is the name of the terraform client id environment variable.
+	TerraformVarClientID = "TF_VAR_CLIENT_ID"
+	//TerraformVarClientSecret is the name of the client secret environment variable.
+	TerraformVarClientSecret = "TF_VAR_CLIENT_SECRET"
+)
+
+// TerraformVariablesEnvirnmentFromClientAuth computes the Terraformer variables environment from the
+// given ServiceAccount.
+func TerraformVariablesEnvirnmentFromClientAuth(auth *ClientAuth) (map[string]string, error) {
```
Comment (rfranzke) [inline controllers/provider-azure/pkg/internal/terraform.go]:
```suggestion
func TerraformVariablesEnvironmentFromClientAuth(auth *ClientAuth) (map[string]string, error) {
```

File: controllers/provider-azure/pkg/internal/infrastructure/infrastructure.go [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+	"context"
+
+	"github.com/gardener/gardener-extensions/controllers/provider-azure/pkg/internal"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+)
+
+// GetClientAuthFromInfrastucture retrieves the ServiceAccount from the Secret referenced in the given Infrastructure.
+func GetClientAuthFromInfrastucture(ctx context.Context, c client.Client, config *extensionsv1alpha1.Infrastructure) (*internal.ClientAuth, error) {
```
Comment (rfranzke) [inline controllers/provider-azure/pkg/internal/infrastructure/infrastructure.go]:
```suggestion
func GetClientAuthFromInfrastructure(ctx context.Context, c client.Client, config *extensionsv1alpha1.Infrastructure) (*internal.ClientAuth, error) {
```

File: controllers/provider-azure/pkg/controller/infrastructure/actuator_delete.go (hunk 2 above)
Comment (zanetworker) [reply controllers/provider-azure/pkg/controller/infrastructure/actuator_delete.go]: I thought maybe in the future we would have to deleted certain infrastructure components like we need in other infrastructure (for cleanup purposes), so thought wouldn't hurt to add the flow now.

**Pull request merged**